validation/esp32-test/README.md

```diff
@@ -0,0 +1,304 @@
+- Certificate-based authentication
+- Client-server TLS communication
+- RustCrypto provider integration with ESP32
+- Embedded system TLS capabilities
+
+## Architecture
+
+The test implements a complete TLS client-server architecture:
+
+1. **TLS Server**: Runs on ESP32, accepts connections, performs TLS handshake
+2. **TLS Client**: Also runs on ESP32, connects to server, sends test messages
+3. **Real Sockets**: Uses actual TCP sockets (not mock implementations)
+4. **Embedded Certificates**: Includes test certificates compiled into the binary
+
+## Features
+
+- ✅ TLS 1.2 and TLS 1.3 support
+- ✅ ECDSA certificate authentication
+- ✅ AES-GCM and ChaCha20-Poly1305 ciphers
+- ✅ Real network socket communication
+- ✅ Embedded certificate validation
+- ✅ ESP32-specific optimizations
+- ✅ Comprehensive logging
+
+## Prerequisites
+
+### Hardware Requirements
+- ESP32 microcontroller (ESP32, ESP32-S2, ESP32-S3, etc.)
+- USB connection for flashing and monitoring
+
+### Software Requirements
+- Rust 1.88.0 or later
+- ESP-IDF development environment
+- ESP32 Rust toolchain
+- OpenSSL (for certificate generation if needed)
+
+### ESP-IDF Setup
+
+1. Install ESP-IDF:
+```bash
+# Using espup (recommended)
+cargo install espup
+espup install
+
+# Or manual installation following ESP-IDF docs
+```
+
+2. Set up the environment:
+```bash
+# Add to your shell profile
+. $HOME/export-esp.sh
+```
+
+3. Install ESP32 Rust toolchain:
+```bash
+rustup target add riscv32imc-esp-espidf
+# or for ESP32-S2/S3
+rustup target add xtensa-esp32-espidf
+```
+
+## Installation
+
+### Clone and Setup
+```bash
+# From the rustls-rustcrypto workspace root
+cd validation/esp32-test
+
+# Build for ESP32
+cargo build --release
+```
+
+### Dependencies
+
+The crate uses the following key dependencies:
+
+- `rustls` 0.23.x - TLS library
+- `rustls-rustcrypto` - RustCrypto provider (workspace)
+- `esp-idf-svc` - ESP32 services
+- `esp-idf-hal` - ESP32 hardware abstraction
+- `anyhow` - Error handling
+- `log` - Logging framework
+
+## Configuration
+
+### Feature Flags
+
+Configure the build with appropriate feature flags:
+
+```toml
+[dependencies.rustls-rustcrypto]
+version = "0.0.2-alpha"
+path = "../../"
+default-features = false
+features = [
+    "aead-chacha20poly1305",
+    "alloc",
+    "der",
+    "ecdsa-p256",
+    "fast",
+    "kx-p256",
+    "pkcs8",
+    "sign-ecdsa-p256",
+    "tls12",
+    "verify-ecdsa-p256",
+    "verify-ecdsa-p256-sha256"
+]
+```
+
+### Build Profiles
+
+The crate includes optimized profiles for ESP32:
+
+```toml
+[profile.release]
+opt-level = "s"  # Optimize for size
+debug = false
+
+[profile.dev]
+debug = true     # Symbols for debugging
+opt-level = "z"  # Optimize for size in debug
+```
+
+## Usage
+
+### Running the Test
+
+1. **Flash to ESP32**:
+```bash
+cargo run --release
+```
+
+2. **Monitor Output**:
+```bash
+# In another terminal
+espmonitor /dev/ttyUSB0  # Adjust port as needed
+```
+
+### Expected Output

```

Comment:
DRY: These are test case specific and should be married in the test cases itself not duplicated in the docs making them more maintenance heavy